mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessageConverter.java

```diff
@@ -0,0 +1,140 @@
+import org.apache.james.mime4j.stream.MimeConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.steveash.guavate.Guavate;
+import com.google.common.base.Preconditions;
+
+class DeletedMessageConverter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DeletedMessageConverter.class);
+
+    private final Clock clock;
+
+    DeletedMessageConverter(Clock clock) {
+        this.clock = clock;
+    }
+
+    DeletedMessage convert(DeletedMessageMetadata deletedMessageMetadata, MailboxMessage mailboxMessage) {
+        Preconditions.checkNotNull(deletedMessageMetadata);
+        Preconditions.checkNotNull(mailboxMessage);
+
+        Optional<Message> mimeMessage = parseMessage(mailboxMessage);
+
+        return DeletedMessage.builder()
+            .messageId(deletedMessageMetadata.getMessageId())
+            .originMailboxes(deletedMessageMetadata.getOwnerMailboxes())
+            .user(retrieveOwner(deletedMessageMetadata))
+            .deliveryDate(retrieveDeliveryDate(mimeMessage, mailboxMessage))
+            .deletionDate(ZonedDateTime.ofInstant(clock.instant(), ZoneOffset.UTC))
+            .sender(retrieveSender(mimeMessage))
+            .recipients(retrieveRecipients(mimeMessage))
+            .hasAttachment(mailboxMessage.getAttachments().iterator().hasNext())
+            .subject(mimeMessage.map(Message::getSubject))
+            .build();
+    }
+
+    private Optional<Message> parseMessage(MailboxMessage mailboxMessage) {
+        DefaultMessageBuilder messageBuilder = new DefaultMessageBuilder();
+        messageBuilder.setMimeEntityConfig(MimeConfig.PERMISSIVE);
+        messageBuilder.setDecodeMonitor(DecodeMonitor.SILENT);
+        try {
+            return Optional.ofNullable(messageBuilder.parseMessage(mailboxMessage.getFullContent()));
+        } catch (IOException e) {
+            LOGGER.warn("Can not parse the message {} cause of {}", mailboxMessage.getUid(), e.getMessage());
+            return Optional.empty();
+        }
+    }
+
+    private User retrieveOwner(DeletedMessageMetadata metadata) {
+        Preconditions.checkNotNull(metadata.getOwner(), "Deleted mail is missing owner");
+        return metadata.getOwner();
+    }
+
+    private ZonedDateTime retrieveDeliveryDate(Optional<Message> mimeMessage, MailboxMessage mailboxMessage) {
+        return mimeMessage.map(Message::getDate)
+            .map(Date::toInstant)
+            .map(instant -> ZonedDateTime.ofInstant(instant, ZoneOffset.UTC))
+            .orElse(ZonedDateTime.ofInstant(mailboxMessage.getInternalDate().toInstant(), ZoneOffset.UTC));
+    }
+
+    private MaybeSender retrieveSender(Optional<Message> mimeMessage) {
+        return mimeMessage
+            .map(Message::getSender)
+            .map(Mailbox::getAddress)
+            .map(MaybeSender::getMailSender)
+            .orElse(MaybeSender.nullSender());
+    }
+
+    private List<MailAddress> retrieveRecipients(Optional<Message> message) {
```

Comment:
You can try with _org.apache.james.jmap.model.Envelope.fromMime4JMessage_

Reply:
It is very good suggestion. IMO, we will append to the vault even if no or error recipients. So that I write new one for this.